ratelimits/limiter.go

```diff
@@ -319,14 +325,28 @@ func (l *Limiter) BatchSpend(ctx context.Context, txns []Transaction) (*Decision
 
 	if batchDecision.allowed {
 		if len(newBuckets) > 0 {
-			err = l.source.BatchSet(ctx, newBuckets)
+			createdBuckets, err := l.source.BatchSetNotExisting(ctx, newBuckets)
 			if err != nil {
 				return nil, fmt.Errorf("batch set for %d keys: %w", len(newBuckets), err)
 			}
+			for k, created := range createdBuckets {
+				if !created {
+					// A bucket was created by another request, fall back to
+					// BatchSet.
+					expiredBuckets[k] = newBuckets[k]
```

Comment:
Don't overwrite the k'th entry of expiredBuckets; I think you want `append` here.

Reply:
I don't believe that we're overwriting here. `prepareBatch()` guarantees that we have no duplicate keys. I'm fine with appending so that it's explicit though.

Reply:
d'oh